ark/codegen.cpp

```diff
@@ -0,0 +1,437 @@
+        //
+        // Distribute tasks to processors.
+        //
+        // A sequence [b, e, s] means the range starts from `b`, ends at
+        // `e - 1`, and the step size is `s`.
+        //
+        // Processor ID sequence: [proc_b, proc_e, proc_s], total `n_procs`
+        // Task ID sequence: [task_b, task_e, task_s], total `n_tasks`
+        //
+        // The distribution starts from the processor ID `proc_cur` and wraps
+        // around (`proc_cur - proc_b` is always a multiple of `proc_s`).
+        // If `task_gran` is 1, the distribution is round-robin; otherwise,
+        // the distribution assigns `task_gran` consequent tasks to each
+        // processor, as long as there are enough tasks.
+        // We distribute tasks from smaller task IDs to larger task IDs.
+        // Therefore, the `t`-th assigned task ID of the processor ID
+        // `(proc_cur + proc_s*p)%n_procs` is (p in range [0, n_procs-1]):
+        //
+        // ```
+        // task_b + task_s*(
+        //     p*task_gran +
+        //     t/task_gran*task_gran*n_procs +
+        //     t%task_gran
+        // )
+        // ```
+        //
+        // where the division is integer division.
+        //
+        // Within a single processor, `n_slots` consequent tasks are
+        // distributed to warps and SRAMs. Specifically, say that
+        // "k-th slot" refers to the set of warps `k * slot_n_warps` ~
+        // `(k+1) * slot_n_warps - 1` and SRAMs `k * slot_n_sram` ~
+        // `(k+1) * slot_n_sram - 1`, then the `t`-th task is assigned to
+        // the `t%n_slots`-th slot.
+        //
+        // Therefore, the `i`-th assigned task ID of the processor ID
+        // `(proc_cur + p)%n_procs` and the `k`-th slot is (p in range
+        // [0, n_procs-1], k in range [0, n_slots-1]) the same as the above
+        // formula with `t` replaced by `k + i*n_slots`:
+        //
+        // ```
+        // task_b + task_s*(
+        //     p*task_gran +
+        //     (k + i*n_slots)/task_gran*task_gran*n_procs +
+        //     (k + i*n_slots)%task_gran
+        // )
+        // ```
+        //
+        // The corresponding CUDA code is generated as follows, saying that
+        // `blockIdx.x` is the processor ID:
+        //
+        // ```
+        // if ((blockIdx.x >= proc_b) &&
+        //     (blockIdx.x < proc_e) &&
+        //     ((blockIdx.x - proc_b) % proc_s == 0)) {
+        //   size_t p = ((blockIdx.x + gridDim.x - proc_cur) % gridDim.x) /
+        //              proc_s;
+        //   size_t k = threadIdx.x / warp_size / slot_n_warps;
+        //   size_t task_id_base = task_b + task_s*p*task_gran;
+        //   for (size_t t = k; ; t += n_slots) {
+        //     size_t task_id = task_id_base + task_s*(
+        //       t/task_gran*task_gran*n_procs + t%task_gran
+        //     );
+        //     if (task_id >= task_e) break;
+        //     task_func(_buf, task_id, sram_per_warp);
+        //   }
+        //   __syncthreads();
```

Comment:
Will we call `__syncthreads()` after each task? Or we will call `__syncthreads()` when task across resourceGroups

Reply:
No, we actually don't call `__syncthreads()` here. I updated comments in [1fbf38b](https://github.com/microsoft/ark/pull/205/commits/1fbf38b4dd839f97199ade30a90959c5145f1a86).